From d60302e6764129b2037136ae746c36f6ac80c38f Mon Sep 17 00:00:00 2001
From: YunQiang Su <yunqiang@isrc.iscas.ac.cn>
Date: Wed, 15 Apr 2026 15:48:55 +0800
Subject: [PATCH 1/2] Support Vendor openruyi

openRuyi is a rolling release focusing on RISC-V support, which is
launched by Institute of Software, CAS.

The default ISA level is set to RVA23U64.

See: https://openruyi.cn/
---
 clang/lib/Driver/ToolChains/Arch/RISCV.cpp    | 4 ++++
 llvm/include/llvm/TargetParser/Triple.h       | 3 ++-
 llvm/include/llvm/TargetParser/TripleName.def | 1 +
 llvm/unittests/TargetParser/TripleTest.cpp    | 6 ++++++
 4 files changed, 13 insertions(+), 1 deletion(-)

diff --git a/clang/lib/Driver/ToolChains/Arch/RISCV.cpp b/clang/lib/Driver/ToolChains/Arch/RISCV.cpp
index 8e650ddf92df..df1695d67d85 100644
--- a/clang/lib/Driver/ToolChains/Arch/RISCV.cpp
+++ b/clang/lib/Driver/ToolChains/Arch/RISCV.cpp
@@ -382,6 +382,8 @@ std::string riscv::getRISCVArch(const llvm::opt::ArgList &Args,
         return "rv64imafdcv_zba_zbb_zbs";
       if (Triple.isOSFuchsia())
         return "rva22u64_v";
+      if (Triple.getVendor() == llvm::Triple::openRuyi)
+        return "rva23u64";
       return "rv64imafdc";
     }
   }
@@ -403,6 +405,8 @@ std::string riscv::getRISCVArch(const llvm::opt::ArgList &Args,
     return "rv64imafdcv_zba_zbb_zbs";
   if (Triple.isOSFuchsia())
     return "rva22u64_v";
+  if (Triple.getVendor() == llvm::Triple::openRuyi)
+    return "rva23u64";
   return "rv64imafdc";
 }
 
diff --git a/llvm/include/llvm/TargetParser/Triple.h b/llvm/include/llvm/TargetParser/Triple.h
index b369b66960b1..6b5b66d0648e 100644
--- a/llvm/include/llvm/TargetParser/Triple.h
+++ b/llvm/include/llvm/TargetParser/Triple.h
@@ -296,7 +296,8 @@ public:
     OpenEmbedded,
     Intel,
     Meta,
-    LastVendorType = Meta
+    openRuyi,
+    LastVendorType = openRuyi
   };
   enum OSType {
     UnknownOS,
diff --git a/llvm/include/llvm/TargetParser/TripleName.def b/llvm/include/llvm/TargetParser/TripleName.def
index 95026bd39db7..ed42efdb9980 100644
--- a/llvm/include/llvm/TargetParser/TripleName.def
+++ b/llvm/include/llvm/TargetParser/TripleName.def
@@ -201,6 +201,7 @@ TRIPLE_VENDOR(SCEI, "scei")
 TRIPLE_VENDOR_ALIAS(SCEI, "sie")
 TRIPLE_VENDOR(SUSE, "suse")
 TRIPLE_VENDOR(Meta, "meta")
+TRIPLE_VENDOR(openRuyi, "openruyi")
 
 #undef TRIPLE_VENDOR
 #undef TRIPLE_VENDOR_ALIAS
diff --git a/llvm/unittests/TargetParser/TripleTest.cpp b/llvm/unittests/TargetParser/TripleTest.cpp
index a81801dc97a7..27ce6385de10 100644
--- a/llvm/unittests/TargetParser/TripleTest.cpp
+++ b/llvm/unittests/TargetParser/TripleTest.cpp
@@ -862,6 +862,12 @@ TEST(TripleTest, ParsedIDs) {
   EXPECT_EQ(Triple::Hurd, T.getOS());
   EXPECT_EQ(Triple::GNU, T.getEnvironment());
 
+  T = Triple("riscv64-openruyi-linux");
+  EXPECT_EQ(Triple::riscv64, T.getArch());
+  EXPECT_EQ(Triple::openRuyi, T.getVendor());
+  EXPECT_EQ(Triple::Linux, T.getOS());
+  EXPECT_EQ(Triple::UnknownEnvironment, T.getEnvironment());
+
   T = Triple("armv7hl-suse-linux-gnueabi");
   EXPECT_EQ(Triple::arm, T.getArch());
   EXPECT_EQ(Triple::SUSE, T.getVendor());
-- 
2.54.0

